### ISSUE TITLE
QUIC connection stalls: one side stops receiving data, other side disconnects

### ISSUE THREAD

**Issue by irealing** [quic-go/quic-go#4704]
**Description:**

During data transmission over a QUIC connection using quic-go, one side of the connection will unexpectedly stop receiving data, while the other side eventually disconnects. The side that stops receiving data remains blocked in a Read operation, even though the peer disconnects, leading to a stalled connection.

This issue occurs in a larger program that involves high-frequency message exchange using QUIC streams. I also see occasional timeout: no recent network activity errors, but the primary issue is the unexpected stall and disconnection.
Steps to Reproduce:

    Establish a QUIC connection between a server and a client over 127.0.0.1.
    Use protobuf to define a Message for sending data and Ack for acknowledgements.
    Continuously send messages from the server to the client, with the client responding with Ack for each message.
    After sending a predefined number of messages, the client marks the final Ack as complete and terminates the connection.
    The issue occurs after some period of normal data exchange, where one side (typically the receiver) stops receiving data, and the sender disconnects after a timeout.

Expected Behavior:

The connection should continue transmitting data until all messages are sent and acknowledged, with no unexpected stalls or disconnections.
Actual Behavior:

    One side (typically the receiver) stops receiving data: It becomes blocked on a Read operation indefinitely, even though the sender is still active or has disconnected.
    The other side (typically the sender) eventually disconnects: It may disconnect after a timeout or due to a failure to receive acknowledgements from the peer.
    Occasionally, I also encounter the error timeout: no recent network activity, but this is not the primary issue.

Configuration:

    quic-go version: v0.48.0 (replace with your version)
    Go version: go1.23.0 (replace with your Go version)
    Operating System: Windows 10

**Code Sample:**

The following is a simplified version of the relevant code. In the full application, the problem can occur under heavier load or after some time.
```go
// Server-side code
func listenAndSend(ctx context.Context) error {
    udpConn, _ := net.ResolveUDPAddr("udp", "127.0.0.1:65533")
    conn, err := net.ListenUDP("udp", udpConn)
    if err != nil {
        return err
    }
    tr := quic.Transport{Conn: conn}
    l, err := tr.Listen(&tls.Config{Certificates: []tls.Certificate{cert}, InsecureSkipVerify: true, ClientAuth: tls.NoClientCert}, quicConfig)
    if err != nil {
        return err
    }
    c, err := l.Accept(ctx)
    if err != nil {
        return err
    }
    defer c.CloseWithError(quic.ApplicationErrorCode(0), "")
    stream, err := c.OpenStreamSync(ctx)
    if err != nil {
        return err
    }
    defer stream.Close()
    reader := &transport{stream: stream}
    m := &Message{Data: make([]byte, 8*1024)}
    ack := &Ack{}
    for {
        select {
        case <-ctx.Done():
            return ctx.Err()
        default:
        }
        _, _ = rand.Read(m.Data)
        _, err = protodelim.MarshalTo(reader, m)
        if err != nil {
            return err
        }
        if err = protodelim.UnmarshalFrom(reader, ack); err != nil {
            return err
        }
        if ack.Complete {
            return nil
        }
    }
}

// Client-side code
func connectAndReceive(ctx context.Context, m int) error {
    conn, err := quic.DialAddr(ctx, "127.0.0.1:65533", &tls.Config{InsecureSkipVerify: true}, quicConfig)
    if err != nil {
        return err
    }
    stream, err := conn.AcceptStream(ctx)
    if err != nil {
        return err
    }
    defer stream.Close()
    defer stream.CancelRead(quic.StreamErrorCode(0))
    defer stream.CancelWrite(quic.StreamErrorCode(0))
    tr := &transport{stream: stream}
    var (
        msg Message
        ack Ack
    )
    i := 0
    for {
        select {
        case <-ctx.Done():
            return ctx.Err()
        default:
        }
        if err = protodelim.UnmarshalFrom(tr, &msg); err != nil {
            return err
        }
        i++
        ack.Ok, ack.Received, ack.Complete = true, int32(len(msg.Data)), i > m
        if _, err = protodelim.MarshalTo(tr, &ack); err != nil {
            return err
        }
        if ack.Complete {
            return nil
        }
    }
}
```
Configuration Used:
```go
var quicConfig = &quic.Config{
    MaxConnectionReceiveWindow: 100 << 20,
    MaxStreamReceiveWindow:     10 << 20,
    MaxIdleTimeout:             time.Second * 30,  // 30 seconds
}
```
Troubleshooting Steps Taken:

    Adjusted flow control windows: Increased MaxConnectionReceiveWindow and MaxStreamReceiveWindow, but the issue persists.
    Extended MaxIdleTimeout: Increased the idle timeout to 1 minute, but the problem still occurs.
    Checked for continuous data flow: Verified that data should be continuously sent (no intentional idle time between messages).
    Simplified the test case: Reduced the complexity to focus on basic message sending and receiving, but the issue remains.

Observations:

    The issue occurs randomly during the message exchange.
    One side (typically the receiver) will stop receiving data and remain blocked in the Read operation, while the other side (the sender) eventually times out and disconnects.
    This happens even though the connection is established over a local 127.0.0.1 address, which should not experience network instability.

**Comment by marten-seemann** [comment]
I recommend enabling qlogs and using qvis to see what's going on. Feel free to open a new issue once you've figured out that there's actually a problem.

**Comment by irealing** [comment]
My apologies. I tried compiling and running the code in a Linux environment, and it worked fine. Even cross-compiling in Linux and then executing the binary in a Windows environment worked without issues.

**Comment by irealing** [comment]
> I recommend enabling qlogs and using qvis to see what's going on. Feel free to open a new issue once you've figured out that there's actually a problem.

My apologies. I tried compiling and running the code in a Linux environment, and it worked fine. Even cross-compiling in Linux and then executing the binary in a Windows environment worked without issues.

**Comment by marten-seemann** [comment]
From what you're describing, this most likely is a flow control issue. In general, you need to make sure that you're reading from a stream. Look out for BLOCKED frames being in the qlog.
